### PR TITLE
feat(kad): expose a kad query facility allowing dynamic num_results

### PR DESCRIPTION
**Pull request by maqi**
## Description

This PR is to expose a kad query facility that allowing specify num_results dynamically.
It is related to the [Sybil Defence issue](https://github.com/libp2p/rust-libp2p/issues/4769),
that during the attempt of implementation on higher level code, it is find will be useful if libp2p-kad can expose such facility.

The PR try not to cause any interference to the existing work flow, only introduce an `extra exposal`. 

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/src/query.rs
Comment (guillaumemichel) [inline protocols/kad/src/query.rs]: Can be consolidated with `continue_iter_closest` by adding `Option<NonZeroUsize>` since the function is only used in this crate.

If no `num_results` is provided, use `self.config.replication_factor`.

File: protocols/kad/src/query.rs
Comment (guillaumemichel) [inline protocols/kad/src/query.rs]: Can be consolidated with `add_iter_closest` by adding a `Option<NonZeroUsize>` parameter. This shouldn't be a problem since the function is only used inside of this crate.

File: protocols/kad/src/behaviour.rs
Comment (guillaumemichel) [inline protocols/kad/src/behaviour.rs]: Alternatively, `num_result` can be included in `QueryInfo::GetClosestPeers` so the signatures of `continue_iter_closest` and `add_iter_closest` wouldn't change.

File: protocols/kad/src/behaviour.rs
Comment (guillaumemichel) [inline protocols/kad/src/behaviour.rs]: I wonder whether it would be better to have a public function `get_closest_peers_with_config` accepting a `Config` including `num_results` instead of `get_closest_peers_num_results`. This would allow to keep a lean interface in the future if more parameters are added to the `get_closest_peers` function?

**Review by guillaumemichel — CHANGES_REQUESTED** [review]
Thanks @maqi for your contribution, this change is indeed needed! I left a few remarks.

File: protocols/kad/src/behaviour.rs
Comment (maqi) [reply protocols/kad/src/behaviour.rs]: So far, the `Config` itself is not exposed outside.
Also, the population of the `Config` might requires access to the `current` config, as the calling place may not have the knowledge of other config variants.

Hence, in this PR, it's being chosen to just pass `num_results` to minimize the change.

**Comment by maqi** [comment]
@guillaumemichel PR updated with using your suggested approach of adding `num_results` into `QueryInfo::GetClosestPeers`, to avoid introduce extra functions.

also replied to your another comment regarding the `with_config` idea, which I think better not to be included in this PR, if really needed ?

btw. any idea why there is a failing CI https://github.com/libp2p/rust-libp2p/actions/runs/10390288352/job/28770402300?pr=5555 complaining `failure enum_struct_variant_field_added: pub enum struct variant field added` ?
and how to resolve it? thx

**Comment by maqi** [comment]
seems the failing CI requires a major version up on libp2p-kad, as a new variant added into the struct.

Though I think it is not necessary, I have to increase the libp2p-kad version to 0.47.0, to get an ALL green CI.

File: protocols/kad/src/behaviour.rs
Comment (guillaumemichel) [inline protocols/kad/src/behaviour.rs]:
```suggestion
        let peer_keys: Vec<kbucket::Key<PeerId>> = self.kbuckets.closest_keys(&target).collect();
        self.queries.add_iter_closest(target, peer_keys, info)
```

File: protocols/kad/src/query.rs
Comment (guillaumemichel) [inline protocols/kad/src/query.rs]:
```suggestion
        let num_results = match info {
            QueryInfo::GetClosestPeers {
                num_results: Some(val),
                ..
            } => val,
            _ => self.config.replication_factor,
        };

        let cfg = ClosestPeersIterConfig {
            num_results,
```

**Comment by maqi** [comment]
Hi, @guillaumemichel 

thx for the review.
PR updated according to your new comments.

**Comment by guillaumemichel** [comment]
Thanks @maqi, it looks good. It would be great to add a test for the cases where `num_result` is higher and lower than `K_VALUE`.

**Comment by maqi** [comment]
Hi, @guillaumemichel 

I have updated the PR to add a test covering the num_results range of K_VALUE / 2 .. K_VALUE * 2

However, it shows the returned result is capped after exceeds K_VALUE .
A quick read into the code shows it might needs a deeper refactor of the code base.

However, for us, a capped result is acceptable, hence I only put the test according to the current impl.
And put some comments to explain the situation.

I think to support `results not capped` , it will be better to have it as a separate PR ?

**Comment by guillaumemichel** [comment]
@maqi thanks for adding the tests!

> However, for us, a capped result is acceptable.

What is your use case? If you need to pass a `num_results` to receive less peers than `K_VALUE` you might as well just take a subset of the returned peers with the default `get_closest_peers`. The feature seems incomplete if it is not possible to increase the number of returned peers to match `num_results` and I am afraid it will be confusing for users.

**Comment by maqi** [comment]
Hi, @guillaumemichel 

For our case, we need to set the `replication_factor` to a value lower than the `K_VALUE`, but still want the `query` of get_closest gives a `definable` num of results ranging from that `replication_factor` to `K_VALUE`.

Or, you mean the `get_closest` will gives `K_VALUE` results anyway (as long as there are enough nodes in the network) ?
but I see the code of [set num_results to replication_factor](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/query.rs#L142) ?

File: protocols/kad/src/behaviour.rs
Comment (guillaumemichel) [inline protocols/kad/src/behaviour.rs]: If the number of returned results is limited by `K_VALUE` let's bound `num_results` to `K_VALUE` to set clear expectations and prevent unexpected behaviour.

**Review by guillaumemichel — COMMENTED** [review]
Alright, I see! A high `replication_factor` or `num_results` is never expected to return more than `K_VALUE` results anyway.

IMO it is fine, because it was never possible to have more than `K_VALUE` closest peers in response, this change simply allows users to define a number of expected closest peers between `1` and `K_VALUE`. Previously, the number of returned closest peers was set to `replication_factor`.

Removing the cap might be tricky, since it would involve forging a new key and additional requests. It can be left to a follow up PR 👍🏻

**Comment by maqi** [comment]
Hi, @guillaumemichel ,  

thx for the summary comment.
PR updated according to your new comment and added some of your word as code comment

**Comment by guillaumemichel** [comment]
Sorry for the back-and-forth, but it would be good to have different values of `replication_factor` in the test. `K_VALUE` (default) and a lower value (e.g `K_VALUE / 2`) should be enough.

**Comment by maqi** [comment]
Hi, @guillaumemichel ,

updated the test according to your suggestion.

**Review by guillaumemichel — APPROVED** [review]
Thanks for your contribution @maqi, it looks good to me!

@jxs could you also have a look before we merge?

File: protocols/kad/src/behaviour.rs
Comment (jxs) [inline protocols/kad/src/behaviour.rs]:
```suggestion
    /// Initiates an iterative query for the closest peers to the given key.
    /// The expected responding peers is specified by `num_results`
    /// Note that the result is capped after exceeds K_VALUE
    ///
    /// The result of the query is delivered in a
    /// [`Event::OutboundQueryProgressed{QueryResult::GetClosestPeers}`].
```

**Comment by maqi** [comment]
Hi, @guillaumemichel ,

may I know the schedule of when the PR will get the second approval to be merged pls?
trying to resolve the conflicts contstantly is little bit annoying  :)

File: protocols/kad/src/behaviour/test.rs
Comment (jxs) [inline protocols/kad/src/behaviour/test.rs]:
```suggestion
// Test the result of get_closest_peers with different num_results
// Note that the result is capped after exceeds K_VALUE
#[test]
```

File: protocols/kad/src/behaviour/test.rs
Comment (jxs) [inline protocols/kad/src/behaviour/test.rs:435]: this function is not used elsewhere right? We can inline its code here

File: protocols/kad/src/behaviour.rs
Comment (jxs) [inline protocols/kad/src/behaviour.rs]:
```suggestion
    pub fn get_n_closest_peers<K>(&mut self, key: K, num_results: NonZeroUsize) -> QueryId
```
wdyt?

**Review by jxs — COMMENTED** [review]
Hi, and thanks for the interest! Left some comments

File: protocols/kad/src/behaviour/test.rs
Comment (maqi) [reply protocols/kad/src/behaviour/test.rs:435]: make it a separate function to avoid the code becomes too `nested`
it's for a better reading experience and coding format.

I'd prefer not to inline the code block here, unless really `have to` ?

**Comment by maqi** [comment]
Hi, @jxs ,

Thx for the review.
PR updated according to your comments.

Replied to one comment, as I'm leading toward `not to inline` the code block.
Let me know if I `have to`, thx.

**Review by jxs — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
